cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaOffsetResolver.java

```diff
@@ -0,0 +1,321 @@
+                                        timeStamp - MAX_TIME_GAP, timeStamp + MAX_TIME_GAP);
+      }
+      if (timeStamp > targetTime) {
+        maxOffset = offset - 1;
+      }
+      if (timeStamp < targetTime) {
+        minOffset = offset + 1;
+      }
+    }
+    return -1;
+  }
+
+  /**
+   * Performs a linear search to find the smallest offset of the message with log event time
+   * equal to {@code targetTime}. Stop searching when the current message has log event time smaller than
+   * {@code minTime} or larger than {@code maxTime}
+   *
+   * @return the matching offset or {@code -1} if not found
+   */
+  private long linearSearchAroundOffset(SimpleConsumer consumer, int partition, String clientName,
+                                        long startOffset, long targetTime, long minTime, long maxTime) {
+    long smallestMatchingOffset = -1;
+    // search from startOffset and backward first
+    long searchOffset = startOffset;
+    while (searchOffset >= 0) {
+      long time = getEventTimeByOffset(consumer, partition, clientName, searchOffset);
+      if (time < minTime) {
+        break;
+      }
+      if (time == targetTime) {
+        smallestMatchingOffset = searchOffset;
+      }
+      searchOffset--;
+    }
+    if (smallestMatchingOffset != -1) {
+      return smallestMatchingOffset;
+    }
+    searchOffset = startOffset;
+    // The latest offset returned by fetchOffsetByTime is 1 larger than the offset of the latest received message
+    long maxOffset = fetchOffsetByTime(consumer, partition, LATEST);
+    // search forward from (startOffset + 1) if no matching offset found when searching backward
+    while (++searchOffset < maxOffset) {
+      long time = getEventTimeByOffset(consumer, partition, clientName, searchOffset);
+      if (time == targetTime) {
+        return searchOffset;
+      }
+      if (time > maxTime) {
+        return -1;
+      }
+    }
+    return -1;
+  }
+
+  private long getEventTimeByOffset(SimpleConsumer consumer, int partition, String clientId, long requestOffset) {
+    FetchRequest req = new FetchRequestBuilder()
+      .clientId(clientId).addFetch(topic, partition, requestOffset, 100000).build();
+    FetchResponse fetchResponse = consumer.fetch(req);
+
+    if (fetchResponse.hasError()) {
+      // Something went wrong!
+      short code = fetchResponse.errorCode(topic, partition);
+      if (code == ErrorMapping.OffsetOutOfRangeCode()) {
+        // We asked for an invalid offset. For simple case ask for the last element to reset
+        LOG.debug("FetchRequest offset: {} out of range.", requestOffset);
+        return -1;
+      }
+      LOG.error("Failed to fetch message for topic: {} partition: {} with error code {}.",
+                topic, partition, code);
+    } else {
+      for (MessageAndOffset messageAndOffset : fetchResponse.messageSet(topic, partition)) {
+        long offset = messageAndOffset.offset();
+        if (offset < requestOffset) {
```

Comment:
Not sure, saw this logic before in Twill `SimpleKafkaConsumer#createFetchedMessages` and somewhere else on the internet. Will double check